core/services/ardupilot_manager/ArduPilotManager.py

```diff
@@ -258,3 +325,42 @@ def remove_endpoints(self, endpoints_to_remove: Set[Endpoint]) -> None:
                 raise
 
         self._update_endpoints()
+
+    def get_available_firmwares(self, vehicle: Vehicle) -> Dict[str, str]:
+        firmwares = {}
+        versions = self.firmware_download.get_available_versions(vehicle, self.current_platform)
+        if not versions:
+            raise NoVersionAvailable(f"Failed to find any version for vehicle {vehicle}.")
+        for version in versions:
+            try:
+                url = self.firmware_download.get_download_url(vehicle, self.current_platform, version)
+                firmwares[version] = url
+            except Exception as error:
+                logger.debug(f"Error fetching URL for version {version} on vehicle {vehicle}: {error}")
+        if not firmwares:
+            raise NoVersionAvailable(f"Failed do get any valid URL for vehicle {vehicle}.")
+        return firmwares
+
+    def install_firmware_from_file(self, firmware_path: pathlib.Path, force_update: bool = False) -> None:
+        try:
+            self.mavlink_manager.stop()
+            # With current "while do" approach used to initiate the Ardupilot firmware, killing the main subprocess
+            # does not kill Ardupilot's firmware instances running.
+            # The 'prune' method is used here to kill them and allow updating the firmware file.
+            self.prune_ardupilot_processes()
```

Comment:
As I'm aware of, we currently don't do this safe check, right?

I agree it's important, but think currently behavior is not being changed here.

Reply:
I don't think this is an issue for the Pixhawks, but it could be for the navigator. I'm not sure what happens if we kill the binary while we have non-centered PWMs to the motors. testing could be enough.

> As I'm aware of, we currently don't do this safe check, right?

Yes, but we also don't have this functionality for the navigator.

Reply:
Agree.
As we talked externally, it seems the correct approach would be for us to send a SIGTERM to the process, and the Ardupilot process to handle this disarm/finish routine.

We are currently using a SIGKILL. Will change.

Reply:
Changed to SIGTERM.
Tests pending.